rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+* `nix flake add <from-flake-ref> <to-flake-ref>`: Add an entry to the
+  local registry. (Recall that this overrides the global registry.)
+  For example,
+
+      # nix flake add nixpkgs github:my-repo/my-nixpkgs
+
+  redirects `nixpkgs` to a different repository. Similarly,
+
+      # nix flake add nixpkgs github:NixOS/nixpkgs/444f22ca892a873f76acd88d5d55bdc24ed08757
+
+  pins `nixpkgs` to a specific revision.
+
+  Note that registries only have an effect on flake references used on
+  the command line or when lock files are generated or updated.
+
+* `nix flake remove <flake-ref>`: Remove an entry from the local
+  registry.
+
+* `nix flake pin <flake-ref>`: Compute a locked flake reference and
+  add it to the local registry. For example:
+
+      # nix flake pin nixpkgs
+
+  will add a mapping like `nixpkgs` ->
+  `github:NixOS/nixpkgs/444f22ca892a873f76acd88d5d55bdc24ed08757` to
+  the local registry.
+
+* `nix flake init`: Create a skeleton `flake.nix` in the current
+  directory.
+
+* `nix flake update`: Recreate the lock file from scratch.
+
+* `nix flake check`: Do some checks on the flake (e.g. check that all
+  `packages` are really packages), then build the `checks`
+  derivations. For example, `nix flake check patchelf` fetches the
+  `patchelf` flake, evaluates it and builds it.
+
+* `nix flake clone`: Do a `git clone` to obtain the source of the
+  specified flake, e.g. `nix flake clone dwarffs` will yield a clone
+  of the `dwarffs` repository.
+
+## Offline use
+
+Since flakes replace channels, we have to make sure that they support
+offline use at least as well. Channels only require network access
+when you do `nix-channel --update`. By contrast, the `nix` command
+will periodically fetch the latest version of the global registry and
+of the top-level flake (e.g. `nix build nixpkgs:hello` may cause it to
+fetch a new version of Nixpkgs). To make sure things work offline:
+
+* It's not a failure if we can't fetch the registry.
+
+* It's not a failure if we can't fetch the latest version of a
+  flake. (For example, `fetchGit` no longer fails if we already have
+  the specified repository and `ref`.)
+
+* Fetched flakes are registered as garbage collector roots, so running
+  the garbage collector while on an airplane will not ruin your day.
+
+* There is an option `--no-net` to explicitly prevent updating the
+  registry or doing other things that need the network (such as
+  substituting). This is the default if there are no configured
+  non-loopback network interfaces.
+
+## CI integration
+
+Flakes are useful for CI systems since `flake.nix` and the lock file
+exactly specify the input repositories. For Hydra this means that you
+no longer need to configure any jobset inputs; just a top-level flake
+reference is enough. For example, the [configuration of the `patchelf`
+jobset](https://hydra.nixos.org/jobset/patchelf/master#tabs-configuration)
+just points to `github:NixOS/patchelf` and doesn't need to specify
+`nixpkgs`.
```

Comment:
Will Hydra be able to push the sources fetched by flake into the binary cache? If I understand it correctly those are fetched at eval-time and only linked into the user's .cache/nix folder.

Reply:
In principle it can do that. Currently they'll only end up in the binary cache if they are a dependency of a derivation.